crates/iceberg/src/expr/visitors/expression_evaluator.rs

```diff
@@ -0,0 +1,819 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use fnv::FnvHashSet;
+
+use crate::{
+    expr::{BoundPredicate, BoundReference},
+    spec::{DataFile, Datum, PrimitiveLiteral, Struct},
+    Error, ErrorKind, Result,
+};
+
+use super::bound_predicate_visitor::{visit, BoundPredicateVisitor};
+
+/// Evaluates a [`DataFile`]'s partition [`Struct`] to check
+/// if the partition tuples match the given [`BoundPredicate`].
+///
+/// Use within [`TableScan`] to prune the list of [`DataFile`]s
+/// that could potentially match the TableScan's filter.
+#[derive(Debug)]
+pub(crate) struct ExpressionEvaluator {
+    /// The provided partition filter.
+    partition_filter: BoundPredicate,
```

Comment:
I think a similar situation applies here to that which you commented on in my PR here: https://github.com/apache/iceberg-rust/pull/347#discussion_r1582148013

If the only property inside the `ExpressionEvaluator` struct is the `BoundPredicate`, we may as well keep the bound predicate as local variables inside the `try_stream!`. Then we only need to have a single struct defined here instead of both `ExpressionEvaluator` and `ExpressionEvaluatorVisitor`, and don't need the `ExpressionEvaluatorCache` either, as we can implement it the same as the proposed `InclusiveMetricsEvaluator` [here](https://github.com/apache/iceberg-rust/pull/347/files#diff-bbfbe5e334be6c501ba2ca0ddd84d658ff0f3f84f2b5b532212f4e096a09e09bR236-R243).

Reply:
@sdd 
I think this won't work since we need the InclusiveProjection for the ExpressionEvaluator; whereas the InclusiveMetricsEvaluator operates on the [row filter](https://github.com/apache/iceberg-python/blob/main/pyiceberg/table/__init__.py#L1695-L1697) of the TableScan which we don't need to cache.

Reply:
OK, makes sense!